File: question/neetcode150/backtracking/subsets-ii.py
from typing import List

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        ans = set()

        s = []

        def dfs(i):
            if i==len(nums):
                # print(tuple(sorted(s)))
                ans.add(tuple(sorted(s)))
                return

            s.append(nums[i])
            dfs(i+1)
            s.pop()
            dfs(i+1)

        dfs(0)
        return [list(s) for s in ans]

File: question/neetcode/backtracking/subsets-ii.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:

    void dfs(int i, int n, vector<int>& cur, vector<vector<int>>& ans, vector<int>& nums){
        if(i==n){
            ans.push_back(cur);
            return;
        }

        cur.push_back(nums[i]);
        dfs(i+1,n,cur,ans,nums);
        cur.pop_back();
        
        while(i+1<n and nums[i+1]==nums[i]) i++;
        dfs(i+1,n,cur,ans,nums);
    }

    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());

        int n = nums.size();
        vector<int> cur;
        vector<vector<int>> ans;

        dfs(0,n,cur,ans,nums);

        return ans;
    }
};